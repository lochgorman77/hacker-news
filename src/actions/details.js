import { HackerNewsApi } from 'api/config';
import { detailsActions } from 'constants/ActionTypes';

export function fetchDetailsSuccess(detail) {
  return {
    type: detailsActions.FETCH_DETAILS_SUCCESS,
    payload: detail,
  };
}
export function fetchDetailsError(error) {
  return {
    type: detailsActions.FETCH_DETAILS_ERROR,
    payload: error,
  };
}
export function fetchDetails(detailId) {
  return function (dispatch) {
    dispatch({ type: detailsActions.FETCH_DETAILS });

    return HackerNewsApi.get(`/item/${detailId}.json`)
      .then((result) => {
        dispatch(fetchDetailsSuccess(result.data));
      })
      .catch((error) => {
        dispatch(fetchDetailsError(error));
      });
  };
}
