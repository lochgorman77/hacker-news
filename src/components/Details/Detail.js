import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DateTime } from 'luxon';
import { withStyles } from '@material-ui/styles';
import { FormattedMessage } from 'react-intl';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';

const styles = (theme) => ({
  root: {
    overflow: 'visible',
    minWidth: 500,
  },
  comment: {
    marginLeft: 'auto',
  },
});

const units = ['year', 'month', 'week', 'day', 'hour', 'minute', 'second'];

export class Detail extends Component {
  timeAgo = (date) => {
    const dateTime = DateTime.fromSeconds(date);
    const diff = dateTime.diffNow().shiftTo(...units);
    const unit = units.find((unit) => diff.get(unit) !== 0) || 'second';

    const relativeFormatter = new Intl.RelativeTimeFormat('en', {
      numeric: 'auto',
    });
    return relativeFormatter.format(Math.trunc(diff.as(unit)), unit);
  };

  render() {
    const { detail, classes } = this.props;

    const time = this.timeAgo(detail.time);
    return (
      <Card className={classes.root} raised>
        <CardHeader title={detail.title} />
        <CardContent>
          <Typography variant="caption" color="textSecondary">
            {detail.score} <FormattedMessage id={'detail.view.by'} /> {detail.by}{' '}
            {time.toLocaleString()}
          </Typography>
        </CardContent>
        <CardActions disableSpacing>
          <Button size="small">
            <Link href={detail.url} color="secondary" target="_blank" rel="noopener">
              <FormattedMessage id={'detail.view.link'} />
            </Link>
          </Button>
          <Button size="small" className={classes.comment}>
            {detail.descendants} <FormattedMessage id={'detail.view.comments'} />
          </Button>
        </CardActions>
      </Card>
    );
  }
}
Detail.propTypes = {
  detail: PropTypes.object,
  classes: PropTypes.object,
};

Detail.defaultProps = {
  detail: {},
};
export const DetailStyle = withStyles(styles, { name: 'Details' })(Detail);
export default DetailStyle;
