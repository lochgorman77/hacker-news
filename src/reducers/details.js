import { detailsActions } from './../constants/ActionTypes';

const initialState = {
  detail: [],
};

export default function roles(state = initialState, action) {
  switch (action.type) {
    case detailsActions.FETCH_DETAILS_SUCCESS:
      return {
        ...state,
        detail: [...state.detail, action.payload],
      };
    case detailsActions.FETCH_DETAILS_ERROR:
      return {
        ...state,
        detail: initialState.detail,
      };
    case detailsActions.RESET:
      return {
        ...state,
        detail: [],
      };
    default:
      return state;
  }
}
export const getDetails = (state) => state.details.detail;
