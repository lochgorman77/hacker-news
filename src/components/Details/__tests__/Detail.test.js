import React from 'react';
import { DetailStyle } from '../Detail';
import details from 'utils/test-data/details-new.json';

const props = { detail: details };

describe('<Detail />', () => {
  let shallow, wrapper;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    wrapper = shallow(<DetailStyle {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper.name()).toBe('MuiCard');
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
