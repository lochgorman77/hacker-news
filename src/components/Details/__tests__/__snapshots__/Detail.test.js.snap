// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail /> renders correctly 1`] = `
<MuiCard
  className="Details-root"
  raised={true}
>
  <MuiCardHeader
    title="Appleâ€™s Anti-Tracking Plans for iPhone"
  />
  <MuiCardContent>
    <MuiTypography
      color="textSecondary"
      variant="caption"
    >
      143
       
      <FormattedMessage
        id="detail.view.by"
      />
       
      nojito
       
      6 hours ago
    </MuiTypography>
  </MuiCardContent>
  <MuiCardActions
    disableSpacing={true}
  >
    <MuiButton
      size="small"
    >
      <MuiLink
        color="secondary"
        href="https://foundation.mozilla.org/en/campaigns/apples-anti-tracking-plans-iphone/"
        rel="noopener"
        target="_blank"
      >
        <FormattedMessage
          id="detail.view.link"
        />
      </MuiLink>
    </MuiButton>
    <MuiButton
      className="Details-comment"
      size="small"
    >
      35
       
      <FormattedMessage
        id="detail.view.comments"
      />
    </MuiButton>
  </MuiCardActions>
</MuiCard>
`;
